timemachine/cpp/src/nonbonded_common.cu

```diff
@@ -0,0 +1,127 @@
+#include <set>
+#include <stdexcept>
+#include <string>
+
+#include "nonbonded_all_pairs.hpp"
+#include "nonbonded_common.cuh"
+
+#include "fanout_summed_potential.hpp"
+#include "summed_potential.hpp"
+
+#include "set_utils.hpp"
+
+namespace timemachine {
+
+void verify_atom_idxs(int N, const std::vector<int> &atom_idxs, const bool allow_empty) {
+    if (atom_idxs.size() == 0) {
+        if (allow_empty) {
+            // No further checks if we allow the indices to be empty
+            return;
+        }
+        throw std::runtime_error("indices can't be empty");
+    }
+    std::set<int> unique_idxs(atom_idxs.begin(), atom_idxs.end());
+    if (unique_idxs.size() != atom_idxs.size()) {
+        throw std::runtime_error("atom indices must be unique");
+    }
+    if (*std::max_element(atom_idxs.begin(), atom_idxs.end()) >= N) {
+        throw std::runtime_error("index values must be less than N(" + std::to_string(N) + ")");
+    }
+    if (*std::min_element(atom_idxs.begin(), atom_idxs.end()) < 0) {
+        throw std::runtime_error("index values must be greater or equal to zero");
+    }
+}
+
+bool is_summed_potential(std::shared_ptr<Potential> pot) {
+    if (std::shared_ptr<FanoutSummedPotential> fanned_potential = std::dynamic_pointer_cast<FanoutSummedPotential>(pot);
+        fanned_potential != nullptr) {
+        return true;
+    } else if (std::shared_ptr<SummedPotential> summed_potential = std::dynamic_pointer_cast<SummedPotential>(pot);
+               summed_potential != nullptr) {
+        return true;
+    }
```

Comment:
unrelated to this PR, but could these runtime type checks be simplified using the [`typeid` operator](https://en.cppreference.com/w/cpp/language/typeid)?